### ISSUE TITLE
Could not find a datatype with identifier xxxx

### ISSUE THREAD

**Issue by HalldorLyngmo** [umbraco/Umbraco.Deploy.Issues#157]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

Multiple v8 versions and v10.3.2

### Bug summary

This is an issue I've only seen on Umbraco Cloud in connection with deployments. Thankfully it's not a frequent issue and it does seem to resolve itself after a couple of deployments in most cases. Nevertheless, it's can be quite frustrating for the customers to watch out for this, especially when it's a baseline/child setup.

It's been there since the beginning of v8 and even though I've not seen it reported on v9 and v10 I was able to reproduce the issue on a v10 project.

In short, there will be a deployment with changes to document and data types where instead of a successful extraction there will be a boot failed or an unsuccessful extraction.

```
[BootFailedException: Boot failed: Umbraco cannot run. See Umbraco's log file for more details. 

-> Umbraco.Core.Exceptions.BootFailedException: Boot failed. 

-> System.ArgumentException: Could not find a datatype with identifier 1163. 
Parameter name: id 
```

The data type can be found in the database and it's not always the same data type which will be the issue.
The workaround is simple. After you restart the site it will boot up just fine and you can extract the changes successfully.

My initial thought was that this was a Deploy issue with extracting changes before Umbraco had completed building the cache on boot-up. Following that, I asked Andy to take a look at a site where the issue could be reproduced consistently. Turns out the failure to boot-up was a red herring. This is what he found:

> I don't think the issue sits within Deploy though.  Here is Deploy is emitting an event saying "I've updated a data type", the CMS responds to that and the exception is thrown within there.  Specifically within updating NuCache.
>
> I tracked down a suspicious-looking TODO within the code path of the stack trace that looks relevant and suggests a form of race condition perhaps causing the issue ([here](https://github.com/umbraco/Umbraco-CMS/blob/v8/dev/src/Umbraco.Web/PublishedCache/NuCache/PublishedSnapshotService.cs#L920) in Umbraco 8, but it's still there in the latest code).

The TODO specifically mentions a potential race condition with data types and content types.

```
 // TODO: need to add a datatype lock
 // this is triggering datatypes reload in the factory, and right after we create some
 // content types by loading them ... there's a race condition here, which would require
 // some locking on datatypes
```
-----
Here is a screenshot from a previous debug session where you can see how many [datatypes ](https://github.com/umbraco/Umbraco-CMS/blob/release-10.3.2/src/Umbraco.Core/Models/PublishedContent/PublishedContentTypeFactory.cs#L67)are in `publishedDataTypes ` when I hit the extraction error and after I restart the site and trigger extraction again.

![image](https://user-images.githubusercontent.com/42831978/211350083-2df2dc52-c6bb-4008-98e3-b2238ff1fffb.png)

### Specifics

```
Exception has been thrown by the target of an invocation. ---> System.ArgumentException: Could not find a datatype with identifier 1134.
Parameter name: id
   at Umbraco.Core.Models.PublishedContent.PublishedContentTypeFactory.GetDataType(Int32 id)
   at Umbraco.Core.Models.PublishedContent.PublishedPropertyType..ctor(String propertyTypeAlias, Int32 dataTypeId, Boolean isUserProperty, ContentVariation variations, PropertyValueConverterCollection propertyValueConverters, IPublishedModelFactory publishedModelFactory, IPublishedContentTypeFactory factory)
   at Umbraco.Core.Models.PublishedContent.PublishedPropertyType..ctor(IPublishedContentType contentType, PropertyType propertyType, PropertyValueConverterCollection propertyValueConverters, IPublishedModelFactory publishedModelFactory, IPublishedContentTypeFactory factory)
   at Umbraco.Core.Models.PublishedContent.PublishedContentTypeFactory.CreatePropertyType(IPublishedContentType contentType, PropertyType propertyType)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Umbraco.Core.Models.PublishedContent.PublishedContentType..ctor(IContentTypeComposition contentType, IPublishedContentTypeFactory factory)
   at Umbraco.Core.Models.PublishedContent.PublishedContentTypeFactory.CreateContentType(IContentTypeComposition contentType)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Umbraco.Web.PublishedCache.NuCache.ContentStore.UpdateDataTypesLocked(IEnumerable`1 dataTypeIds, Func`2 getContentType)
   at Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService.Notify(JsonPayload[] payloads)
   at Umbraco.Core.PublishedModelFactoryExtensions.WithSafeLiveFactoryReset(IPublishedModelFactory factory, Action action)
   at Umbraco.Web.Cache.DataTypeCacheRefresher.Refresh(JsonPayload[] payloads)
   at Umbraco.Core.Sync.ServerMessengerBase.Deliver[TPayload](ICacheRefresher refresher, TPayload[] payload)
   at Umbraco.Web.Cache.DistributedCache.RefreshByPayload[TPayload](Guid refresherGuid, TPayload[] payload)
   at Umbraco.Web.Cache.DistributedCacheExtensions.RefreshDataTypeCache(DistributedCache dc, IDataType dataType)
   at Umbraco.Web.Cache.DistributedCacheBinder.DataTypeService_Saved(IDataTypeService sender, SaveEventArgs`1 e)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Umbraco.Web.Cache.DistributedCacheBinder.HandleEvents(IEnumerable`1 events)
   at Umbraco.Deploy.DeployEventDispatcher.ScopeExitCompleted()
   at Umbraco.Core.Events.QueuingEventDispatcherBase.ScopeExit(Boolean completed)
   at Umbraco.Core.Scoping.Scope.<>c__DisplayClass85_0.<RobustExit>b__1()
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)
   at Umbraco.Core.Scoping.Scope.RobustExit(Boolean completed, Boolean onException)
   at Umbraco.Core.Scoping.Scope.Dispose()
   at Umbraco.Deploy.Work.WorkItems.DiskReadWorkItem.<ExecuteAsyncSub>d__16.MoveNext()
```

### Steps to reproduce

I don't have a way to reproduce it without using Deploy and UDA files I got from a customer project. 
The method would be to:
1. Create a v8 project (or v10) on Umbraco Cloud. 
2. Extract one set of UDA files to the environment.
3. Copy modified UDA files to the environment.
4. Trigger an extraction and the exception should pop up in the logs. 

Reach out to me and I'll ship you the two sets of UDA files.

### Expected result / actual result

I expect a successful extraction and the Nucache to be updated without any issues.

---
_This item has been added to our backlog [AB#25490](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/25490)_

**Comment by ronaldbarendse** [comment]
I've been debugging this issue and can confirm it is indeed triggered by Deploy saving multiple data types in a single scope (as part of a schema deployment), but only when one of those is a newly added data type. However, the exception is caused by the CMS not correctly updating an [internal `PublishedDataType` lookup dictionary in `PublishedContentTypeFactory`](https://github.com/umbraco/Umbraco-CMS/blob/v8/dev/src/Umbraco.Core/Models/PublishedContent/PublishedContentTypeFactory.cs#L100-L120):
- This lookup dictionary is populated with all data types on the first run (hence it works again after a restart), but once that's done, only the data types specified in the [`NotifyDataTypeChanges()` call will be updated](https://github.com/umbraco/Umbraco-CMS/blob/v8/dev/src/Umbraco.Web/PublishedCache/NuCache/PublishedSnapshotService.cs#L920-L924);
- After saving a data type, the published content cache (NuCache) is rebuild, which recreates the `PublishedContentType` that requires all data types used on that content type to exist in this internal lookup dictionary;
- If multiple data types are saved in a single scope, multiple rebuilds are done (the CMS already avoids this for content, media and member types [by grouping the events](https://github.com/umbraco/Umbraco-CMS/blob/v8/dev/src/Umbraco.Web/Cache/DistributedCacheBinder.cs#L75-L191));
- If the first saved data type requires updating a content/media/member type that also contains the other newly added data type, the internal lookup dictionary won't contain that new data type and throw this exception...

TL;DR: The CMS currently doesn't correctly handle saving multiple data types when creating a `PublishedContentType` as part of rebuilding the published content cache (NuCache), resulting in the `Could not find a datatype with identifier...` exception.

-------------------

There are some differences when data types are saved as part of a schema deployment in Deploy (compared to when saved in the back-office), making the exception more likely to occur:
- A schema deployment can easily contain multiple changes to data types in a single scope, while changes done in the back-office will create separate scopes for each operation/request;
- Deploy supresses all events by using a custom `IEventDispatcher` when creating a new scope, ensuring the handlers aren't raised/invoked after completing the scope, but the `IDistributedCacheBinder` does handle the events to ensure the cache refreshers are notified and updated correctly;
- All events are handled, compared to the CMS that only handles the ['last in' events in the `QueuingEventDispatcher`](https://github.com/umbraco/Umbraco-CMS/blob/v8/dev/src/Umbraco.Core/Events/QueuingEventDispatcher.cs#L23) - I'm hesitant to update Deploy to do the same, but this could reduce the amount of (duplicate) events that are handled;
- Because all of this, multiple `DataTypeCacheRefresher` payloads are sent, triggering multiple rebuilds of the published content cache (NuCache).

We'll update Deploy v4+ to group the data type saved events/notifications as @nikolajlauridsen also suggested, so users can easily update to the latest Deploy version. I'll also create a PR for the CMS to do the same, as this can use most of the existing event grouping that's already implemented for content/media/member types 👍🏻

**Comment by AndyButland** [comment]
This fix will be released in the next patch releases for Deploy, due next Tuesday.
